mldsa/sign.c

```diff
@@ -143,71 +144,80 @@ int crypto_sign_signature_internal(uint8_t *sig, size_t *siglen,
   polyveck_ntt(&s2);
   polyveck_ntt(&t0);
 
-rej:
-  /* Sample intermediate vector y */
-  polyvecl_uniform_gamma1(&y, rhoprime, nonce++);
+  /* Reference: This code is re-structured using a while(true), */
+  /* with explicit "continue" statements rather than the "goto" */
+  /* so there is a syntactic location to place the loop         */
+  /* invariants for CBMC.                                       */
+  while (true)
+  __loop__(
+    invariant(true); /* TODO */
+  )
+  {
+    /* Sample intermediate vector y */
+    polyvecl_uniform_gamma1(&y, rhoprime, nonce++);
 
-  /* Matrix-vector multiplication */
-  z = y;
-  polyvecl_ntt(&z);
-  polyvec_matrix_pointwise_montgomery(&w1, mat, &z);
-  polyveck_reduce(&w1);
-  polyveck_invntt_tomont(&w1);
+    /* Matrix-vector multiplication */
+    z = y;
+    polyvecl_ntt(&z);
+    polyvec_matrix_pointwise_montgomery(&w1, mat, &z);
+    polyveck_reduce(&w1);
+    polyveck_invntt_tomont(&w1);
 
-  /* Decompose w and call the random oracle */
-  polyveck_caddq(&w1);
-  polyveck_decompose(&w1, &w0, &w1);
-  polyveck_pack_w1(sig, &w1);
+    /* Decompose w and call the random oracle */
+    polyveck_caddq(&w1);
+    polyveck_decompose(&w1, &w0, &w1);
+    polyveck_pack_w1(sig, &w1);
 
-  shake256_init(&state);
-  shake256_absorb(&state, mu, MLDSA_CRHBYTES);
-  shake256_absorb(&state, sig, MLDSA_K * MLDSA_POLYW1_PACKEDBYTES);
-  shake256_finalize(&state);
-  shake256_squeeze(sig, MLDSA_CTILDEBYTES, &state);
-  poly_challenge(&cp, sig);
-  poly_ntt(&cp);
+    shake256_init(&state);
+    shake256_absorb(&state, mu, MLDSA_CRHBYTES);
+    shake256_absorb(&state, sig, MLDSA_K * MLDSA_POLYW1_PACKEDBYTES);
+    shake256_finalize(&state);
+    shake256_squeeze(sig, MLDSA_CTILDEBYTES, &state);
+    poly_challenge(&cp, sig);
+    poly_ntt(&cp);
+
+    /* Compute z, reject if it reveals secret */
+    polyvecl_pointwise_poly_montgomery(&z, &cp, &s1);
+    polyvecl_invntt_tomont(&z);
+    polyvecl_add(&z, &z, &y);
+    polyvecl_reduce(&z);
+    if (polyvecl_chknorm(&z, MLDSA_GAMMA1 - MLDSA_BETA))
+    {
+      continue;
```

Comment:
```suggestion
      /* reject */
      continue;
```

Similar for other continue statements.

Reply:
Done